src/ruby/puppetserver-lib/puppet/server/compiler.rb

```diff
@@ -14,11 +15,59 @@ def compile(request_data)
         node = create_node(processed_hash)
 
         catalog = Puppet::Parser::Compiler.compile(node, processed_hash['job_id'])
+
+        maybe_save(processed_hash, node.facts, catalog)
+
         catalog.to_data_hash
       end
 
       private
 
+      def maybe_save(processed_hash, facts, catalog)
+        nodename = processed_hash['certname']
+        environment = processed_hash['environment']
+        persist = processed_hash['persistence']
+
+        Puppet.warning(facts.inspect)
+        if persist['facts']
+          save_artifact(:facts, facts, nodename, environment)
+        end
+
+        if persist['catalog']
+          # TODO: We need to pass transaction_id in here some how.
```

Comment:
I think this is my remaining hard todo that I know of (though I should do another round of manual testing). Did you find anything out about transaction_ids @Magisus , I vaguely recall you mentioning the three different ids we need to work with yesterday.

Reply:
Okay, looked it up in the [docs](https://puppet.com/docs/puppet/6.2/http_api/http_catalog.html) (which I think you linked to yesterday). Looks like transaction_uuid can be provided by the http api caller. This may be unnecessary if not being persisted? If provided by the caller then it should be given to the compiler, saved with the puppetdb catalog terminus, and should be submitted back to the report processor. Does that sound correct?

Reply:
Also, does that mean we should update the http body schema to allow optional keys listed in the linked documentation? Sorry if that's something we already talked about (I remembered my schema updates to just be around removing reports from persistence).

Reply:
I plumbed transaction_uuid through the saving, though I haven't taken on any of the other api parameters in those docs.